### PR TITLE
Ensure writable frames

### PR DESCRIPTION
**Pull request by jakirkham**
Fixes https://github.com/dask/distributed/issues/1978
Fixes https://github.com/dask/distributed/issues/3943

Closes https://github.com/dask/distributed/pull/3918 (superseded)

User code often expects NumPy and Pandas objects to be writable. However this fails in various cases. After some investigation this appears to be due to some frames being read-only (as they coerced to `bytes` via compression, logic in `merge_frames`, etc.). To fix this we update the logic in `merge_frames` to check frames are writable. If frames are read-only, we copy them to a writable object (like `bytearray`). We include tests of `merge_frames` and NumPy array serialization to make sure this is working correctly.

### CODE REVIEW

**Comment by mrocklin** [comment]
Thanks for handling this @jakirkham .  I'm only seeing tests in this PR so far.  Is that correct?

**Comment by jakirkham** [comment]
Yeah I'm just trying to demonstrate the failure on CI. Will push the rest of the fix after.

**Comment by jakirkham** [comment]
Alright have pushed the changes to fix the tests. Please let me know what you think 🙂

File: distributed/protocol/utils.py [hunk 1]
```diff
@@ -58,34 +58,36 @@ def merge_frames(header, frames):
...
-                l = 0
-        if len(L) == 1:  # no work necessary
-            out.extend(L)
-        else:
-            out.append(b"".join(L))
-    return out
+                out.append(memoryview(bytearray().join(L)))
+        frames = out
+
+    frames = [memoryview(bytearray(f)) if f.readonly else f for f in frames]
```
Comment (mrocklin) [inline distributed/protocol/utils.py:88]: For fun I decided to micro-benchmark this: 

```python
In [1]: L = [b'0' * i for i in [10, 100, 10, 100, 1000] * 2]                                                                                                                                                      

In [2]: %timeit b"".join(L)                                                                                                                                                                                       
203 ns ± 1.6 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [3]: %timeit memoryview(bytearray().join(L))                                                                                                                                                                   
463 ns ± 5.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

260ns doesn't really matter today I don't think

**Comment by mrocklin** [comment]
In principle this looks good to me.  Thank you for identifying and resolving this @jakirkham !

**Comment by jakirkham** [comment]
Thanks Matt! 😄

**Comment by jakirkham** [comment]
Adding a test for a Pandas Series as well in PR ( https://github.com/dask/distributed/pull/3995 ).

**Pull request merged**